app/services/post_alerter.rb

```diff
@@ -539,6 +546,49 @@ def expand_group_mentions(groups, post)
 
   end
 
+  def expand_here_mention(post, exclude_ids: nil)
+    exclude_ids ||= []
+    user_ids = []
+
+    # users that liked posts in topic
+    if user_ids.size < SiteSetting.max_here_mentioned
+      user_ids += PostAction
+        .where(post_id: Post.select(:id).where(topic_id: post.topic_id))
+        .where(post_action_type_id: PostActionType.types[:like])
+        .group(:user_id)
+        .order('COUNT(user_id) DESC')
+        .pluck(:user_id)
+      user_ids = (user_ids - exclude_ids).uniq
+    end
+
+    # users that read the topic
+    if user_ids.size < SiteSetting.max_here_mentioned
+      user_ids += TopicUser
+        .where(topic_id: post.topic_id)
+        .order(total_msecs_viewed: :desc)
+        .pluck(:user_id)
+      user_ids = (user_ids - exclude_ids).uniq
+    end
+
+    # users that were allowed to topic
+    if user_ids.size < SiteSetting.max_here_mentioned
+      user_ids += TopicAllowedUser
+        .where(topic_id: post.topic_id)
+        .pluck(:user_id)
+      user_ids = (user_ids - exclude_ids).uniq
+    end
+
+    # users that are members of groups allowed to topic
+    if user_ids.size < SiteSetting.max_here_mentioned
+      user_ids += GroupUser
+        .where(group_id: TopicAllowedGroup.select(:group_id).where(topic_id: post.topic_id))
+        .pluck(:user_id)
+      user_ids = (user_ids - exclude_ids).uniq
+    end
+
+    yield User.real.where(id: user_ids[0..SiteSetting.max_here_mentioned])
```

Comment:
I find using `yield` at the end a little odd because we can simply return the results to the caller.